core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileSystem.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file.cache;
+
+import alluxio.AlluxioURI;
+import alluxio.client.file.DelegatingFileSystem;
+import alluxio.client.file.FileInStream;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.URIStatus;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.exception.AlluxioException;
+import alluxio.grpc.OpenFilePOptions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * A FileSystem implementation with a local cache.
+ */
+public class LocalCacheFileSystem extends DelegatingFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalCacheFileSystem.class);
+  private static Optional<CacheManager> sCacheManager;
+
+  private final AlluxioConfiguration mConf;
+
+  /**
+   * @param fs a FileSystem instance to query on local cache miss
+   * @param conf the configuration, only respected for the first call
+   */
+  @edu.umd.cs.findbugs.annotations.SuppressFBWarnings(
+      value = "ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD",
+      justification = "write to static is made threadsafe")
+  public LocalCacheFileSystem(FileSystem fs, AlluxioConfiguration conf) {
+    super(fs);
+    // TODO(feng): support multiple cache managers
+    if (sCacheManager == null) {
+      synchronized (LocalCacheFileSystem.class) {
```

Comment:
Would this be easier if using something like `AtomicReference`?

Reply:
Are you suggesting to use `getAndSet`, that requires creating a new instance of cachemanager each time?